### PR TITLE
Raise an error, and fail execution when a sub-command fails

### PR DESCRIPTION
**Pull request by orta**
# What Changed

# Why

Some parts of our deploy (like maybe the `git push`) would fail, but we'd still get a green build - e.g. https://circleci.com/gh/artsy/reaction/21449 where it's printing out that something critical has occurred but didn't set the exit code correctly and CI didn't register we had a problem. 

I'd rather err on the side of bailing too early with a full throw in these sub-commands (but you have more experience with this, so maybe some are allowed to fail?)

I had to switch from `exec` to `spawn` to reliably get the exit codes, and handle all error cases, which changed the public api for `execSpawn` to go from `"echo 'one string'"` to `"echo", ["args"]` to match the way that API works.

Todo:

- [x] Add tests
- [x] Add docs
- [x] Add yourself to contributors

### CODE REVIEW

File: src/utils/exec-promise.ts [hunk 1]
```diff
@@ -1,14 +1,52 @@
...
-const exec = promisify(childProcess.exec);
+/**
+ * Wraps up running a command into a single promise,
+ * returning the stdout as a string if the command succeeds
+ * and throwing if it does not.
+ *
+ * @param cmd the command as a string to pass in
+ */
+export default async function execPromise(cmd: string, args?: string[]) {
+  return new Promise<string>((completed, reject) => {
```
Comment (Aghassi) [inline src/utils/exec-promise.ts]: May I recommend https://github.com/sindresorhus/execa for this sort of thing? We don't have to handle the logic and can rely on a good open source library to get it right.

**Comment by hipstersmoothie** [comment]
I'll take a look at the commit thing. I do not intend to steal them all! Are you running auto command locally to test them out, setting me as the author?

here is where we are setting the git user https://github.com/intuit/auto-release/blob/master/src/main.ts#L464

it also only runs when you do either changelog, release, or shipit. I would really like to fix this issue as it is annoying me too

File: src/utils/exec-promise.ts (hunk 1 above)
Comment (hipstersmoothie) [reply src/utils/exec-promise.ts]: Yeah I'm down with that

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -6,4 +6,4 @@ Todo:
 
 - [ ] Add tests
 - [ ] Add docs
-- [ ] Add yourself to contributors
+- [ ] Add yourself to contributors (run `yarn contributors:add`)
```
Comment (hipstersmoothie) [inline .github/PULL_REQUEST_TEMPLATE.md]: 👍

File: src/utils/exec-promise.ts (hunk 1 above)
Comment (hipstersmoothie) [reply src/utils/exec-promise.ts]: but that can be a separate PR. Just get the tests passing on this one + rewrite the commits to @orta and I'll merge

**Comment by hipstersmoothie** [comment]
Maybe we could use https://www.npmjs.com/package/is-ci to detect we are in the CI and only set the git user if that's true

**Comment by hipstersmoothie** [comment]
I've open #147 to handle this. will get to it today

**Comment by hipstersmoothie** [comment]
I bet it was the tests that were doing this. Just a heads up @orta when we go to V2 the only breaking changes we made are https://github.com/intuit/auto-release/pull/138

**Comment by hipstersmoothie** [comment]
commit issue should be fixed via #151

**Comment by hipstersmoothie** [comment]
This PR will release v2.0.0

**Comment by orta** [comment]
Alrighty, I've updated this and gonna read feedback now

**Comment by orta** [comment]
OK, this should be 👍

**Comment by zephraph** [comment]
@hipstersmoothie I think we're good when you get a chance to take a peek! 🎉 2.0 hype train.

**Comment by hipstersmoothie** [comment]
Unfortunately this seems to not work 😢 https://circleci.com/gh/intuit/auto-release/1608

**Pull request merged**